tests/test_body_match.js

```diff
@@ -1,110 +1,84 @@
+    body: JSON.stringify({ bar: 'foo', foo: 'bar' }),
+  })
+
+  t.equal(body, 'Heyyyy!')
+  scope.done()
 })
 
-test('allow unordered body with form encoding', { only: true }, t => {
+test('match form body reagardless of field ordering', async t => {
   const scope = nock('http://example.test')
-    .post('/like-wtf', {
-      foo: 'bar',
-      bar: 'foo',
-    })
+    .post('/', { foo: 'bar', bar: 'foo' })
     .reply(200, 'Heyyyy!')
 
-  mikealRequest(
-    {
-      uri: 'http://example.test/like-wtf',
-      method: 'POST',
-      form: {
-        bar: 'foo',
-        foo: 'bar',
-      },
-    },
-    function(e, r, body) {
-      t.equal(body, 'Heyyyy!')
-      scope.done()
-      t.end()
-    }
-  )
-})
+  const { body } = await got.post('http://example.test/', {
+    form: true,
+    body: { bar: 'foo', foo: 'bar' },
+  })
 
-test('allow string json spec', t => {
-  const bodyObject = { bar: 'foo', foo: 'bar' }
+  t.equal(body, 'Heyyyy!')
+  scope.done()
+})
 
+test('match json body specified as json string', async t => {
   const scope = nock('http://example.test')
-    .post('/like-wtf', JSON.stringify(bodyObject))
+    .post('/', JSON.stringify({ bar: 'foo', foo: 'bar' }))
     .reply(200, 'Heyyyy!')
 
-  mikealRequest(
-    {
-      uri: 'http://example.test/like-wtf',
-      method: 'POST',
-      json: {
-        bar: 'foo',
-        foo: 'bar',
-      },
-    },
-    function(e, r, body) {
-      t.equal(body, 'Heyyyy!')
-      scope.done()
-      t.end()
-    }
-  )
+  const { body } = await got.post('http://example.test/', {
+    body: JSON.stringify({ bar: 'foo', foo: 'bar' }),
+  })
+
+  t.equal(body, 'Heyyyy!')
+  scope.done()
 })
 
-test('match body is regex trying to match string', function(t) {
-  nock('http://example.test')
-    .post('/', new RegExp('a.+'))
-    .reply(200)
+test('match body is regex trying to match string (matches)', async t => {
+  const scope = nock('http://example.test')
+    .post('/', new RegExp('abc'))
+    .reply(201)
 
-  mikealRequest(
-    {
-      url: 'http://example.test/',
-      method: 'post',
-      json: {
-        auth: {
-          passwd: 'abc',
-        },
-      },
-    },
-    function(err, res) {
-      if (err) throw err
-      assert.equal(res.statusCode, 200)
-      t.end()
-    }
-  )
+  const { statusCode } = await got.post('http://example.test/', {
+    body: JSON.stringify({ nested: { value: 'abc' } }),
+  })
+
+  t.is(statusCode, 201)
+  scope.done()
+})
+
+test('match body is regex trying to match string (does not match)', async t => {
+  const scope1 = nock('http://example.test')
+    .post('/', new RegExp('def'))
+    .reply(201)
+  const scope2 = nock('http://example.test')
+    .post('/', new RegExp('.'))
+    .reply(202)
+
+  const { statusCode } = await got.post('http://example.test/', {
+    body: JSON.stringify({ nested: { value: 'abc' } }),
+  })
+
+  t.is(statusCode, 202)
+  t.equal(scope1.isDone(), false)
```

Comment:
I like this, I wish we had more tests for not-matching.

Reply:
Agreed! There are many places where tests like this one should be added. I try to do this when adding new behavior.

I'd really like to hit 100% coverage, mostly so we can avoid creating new untested code paths, though it's clear from examples like these – and the bugs that come up around them – that 100% is only the starting point.